File: Final/worker/worker.py
from flask import Flask, request, jsonify
import os
import time
import requests
import threading
import random

app = Flask(__name__)

# Inicialización de variables
task_queue = []
is_leader = False
leader_url = None
heartbeat_interval = 5  # Intervalo de tiempo en segundos para enviar heartbeats

worker_id = os.getenv('WORKER_ID', 'unknown')
worker_port = os.getenv('PORT', '5000')
hostname = os.getenv('HOSTNAME', 'localhost')
workers = ['http://worker1:5001', 'http://worker2:5002']

def process_task(task):
    print(f"Processing task: {task['task_type']} for client {task['client_id']}")
    file_content = task['file_content']
    task_type = task['task_type']
    keyword = task.get('keyword')
    n = task.get('n')
    result = None
    if task_type == 'word_count':
        result = file_content.lower().count(keyword.lower()) if keyword else len(file_content.split())
    elif task_type == 'keyword_search':
        result = keyword.lower() in file_content.lower()
    elif task_type == 'keyword_repetition':
        result = file_content.lower().split().count(keyword.lower()) >= n
    else:
        result = "Invalid task type"
    print(f"Result for client {task['client_id']} is {result}")
    return result

@app.route('/submit_task', methods=['POST'])
def submit_task():
    try:
        task = request.json
        print(f"Received task: {task}")
        result = process_task(task)
        return jsonify({"status": "Task processed successfully", "result": result}), 200
    except Exception as e:
        print(f"Error processing task: {e}")
        return jsonify({"status": f"Error processing task: {e}"}), 400

@app.route('/heartbeat', methods=['POST'])
def heartbeat():
    global is_leader, leader_url
    try:
        leader_id = request.json['worker_id']
        print(f"Received heartbeat from worker {leader_id}")
        if leader_id != worker_id:
            is_leader = False
            leader_url = f"http://worker{leader_id}:{os.getenv('WORKER_PORT')}"
        else:
            print(f"Worker {worker_id} is the leader")
        return jsonify({"status": "Heartbeat received"}), 200
    except Exception as e:
        print(f"Error receiving heartbeat: {e}")
        return jsonify({"status": f"Error receiving heartbeat: {e}"}), 400

@app.route('/is_leader', methods=['GET'])
def get_leader_status():
    return jsonify({"is_leader": is_leader, "worker_id": worker_id}), 200

def send_heartbeat():
    global leader_url, is_leader
    while True:
        time.sleep(heartbeat_interval)
        try:
            if leader_url and leader_url != f"http://{hostname}:{worker_port}":
                response = requests.post(f"{leader_url}/heartbeat", json={"worker_id": worker_id})
                if response.status_code == 200:
                    print(f"Heartbeat sent from worker {worker_id} to leader {leader_url}")
                else:
                    print(f"Failed to send heartbeat from worker {worker_id} to leader {leader_url}")
            else:
                # Check if current worker should become leader
                print(f"No leader detected, worker {worker_id} is checking for leadership")
                for worker in workers:
                    if worker != f"http://{hostname}:{worker_port}":
                        try:
                            response = requests.get(f"{worker}/is_leader")
                            if response.status_code == 200:
                                data = response.json()
                                if data['is_leader']:
                                    leader_url = worker
                                    break
                        except Exception as e:
                            print(f"Error checking leader status from {worker}: {e}")
                else:
                    is_leader = True
                    leader_url = f"http://{hostname}:{worker_port}"
                    print(f"Worker {worker_id} assuming leadership")
        except Exception as e:
            print(f"Error sending heartbeat from worker {worker_id}: {e}")
            if is_leader:
                leader_url = None
                is_leader = False

if __name__ == '__main__':
    if worker_id == '1':
        is_leader = True
        leader_url = f"http://{hostname}:{worker_port}"
        print(f"Worker {worker_id} assuming leadership")
    threading.Thread(target=send_heartbeat).start()
    
    app.run(host='0.0.0.0', port=int(worker_port))

File: Final/worker-node/worker.js
const express = require('express');
const bodyParser = require('body-parser');
const axios = require('axios');

const app = express();
app.use(bodyParser.json());

const workerId = process.env.WORKER_ID || 'unknown';
const workerPort = process.env.PORT || 5000;
const hostname = process.env.HOSTNAME || 'localhost';
let workers = ["http://worker1:5001", "http://worker2:5002", "http://worker3:5003"];

let isLeader = false;
let leaderUrl = null;
const heartbeatInterval = 1000; // Intervalo de tiempo en milisegundos para enviar heartbeats

const processTask = (task) => {
    console.log(`Processing task: ${task.task_type} for client ${task.client_id}`);
    const fileContent = task.file_content;
    const taskType = task.task_type;
    const keyword = task.keyword;
    const n = task.n;
    let result = null;

    if (taskType === 'word_count') {
        result = keyword ? (fileContent.match(new RegExp(keyword, 'gi')) || []).length : fileContent.split(/\s+/).length;
    } else if (taskType === 'keyword_search') {
        result = fileContent.toLowerCase().includes(keyword.toLowerCase());
    } else if (taskType === 'keyword_repetition') {
        result = (fileContent.match(new RegExp(keyword, 'gi')) || []).length >= n;
    } else {
        result = "Invalid task type";
    }

    console.log(`Result for client ${task.client_id} is ${result}`);
    return result;
};

app.post('/submit_task', (req, res) => {
    try {
        const task = req.body;
        console.log(`Received task: ${JSON.stringify(task)}`);
        const result = processTask(task);
        res.json({ status: "Task processed successfully", result });
    } catch (e) {
        console.error(`Error processing task: ${e}`);
        res.status(400).json({ status: `Error processing task: ${e}` });
    }
});

app.post('/heartbeat', (req, res) => {
    try {
        const leaderId = req.body.worker_id;
        console.log(`Received heartbeat from worker ${leaderId}`);
        if (leaderId !== workerId) {
            isLeader = false;
            leaderUrl = `http://${leaderId}:${workerPort}`;
        } else {
            console.log(`Worker ${workerId} is the leader`);
        }
        res.json({ status: "Heartbeat received" });
    } catch (e) {
        console.error(`Error receiving heartbeat: ${e}`);
        res.status(400).json({ status: `Error receiving heartbeat: ${e}` });
    }
});

app.get('/is_leader', (req, res) => {
    res.json({ is_leader: isLeader, worker_id: workerId });
});

const sendHeartbeat = async () => {
    while (true) {
        try {
            await new Promise(resolve => setTimeout(resolve, heartbeatInterval));
            if (leaderUrl && leaderUrl !== `http://${hostname}:${workerPort}`) {
                try {
                    const response = await axios.post(`${leaderUrl}/heartbeat`, { worker_id: workerId });
                    console.log(`Heartbeat sent from worker ${workerId} to leader ${leaderUrl}`);
                } catch (e) {
                    console.error(`Failed to send heartbeat from worker ${workerId} to leader ${leaderUrl}: ${e}`);
                }
            } else {
                console.log(`No leader detected, worker ${workerId} is checking for leadership`);
                let leaderFound = false;
                for (const worker of workers) {
                    if (worker !== `http://${hostname}:${workerPort}`) {
                        try {
                            const response = await axios.get(`${worker}/is_leader`);
                            if (response.data.is_leader) {
                                leaderUrl = worker;
                                leaderFound = true;
                                break;
                            }
                        } catch (e) {
                            console.error(`Error checking leader status from ${worker}: ${e}`);
                        }
                    }
                }
                if (!leaderFound) {
                    isLeader = true;
                    leaderUrl = `http://${hostname}:${workerPort}`;
                    console.log(`Worker ${workerId} assuming leadership`);
                }
            }
        } catch (e) {
            console.error(`Error sending heartbeat from worker ${workerId}: ${e}`);
            if (isLeader) {
                leaderUrl = null;
                isLeader = false;
            }
        }
    }
};

if (workerId === '1') {
    isLeader = true;
    leaderUrl = `http://${hostname}:${workerPort}`;
    console.log(`Worker ${workerId} assuming leadership`);
}

app.listen(workerPort, () => {
    console.log(`Worker ${workerId} running on port ${workerPort}`);
    sendHeartbeat();
});
